### PR TITLE
DRILL-7203: Accept impersonation userName as form field & fix back button for query page

### PR DESCRIPTION
**Pull request by dobesv**
# [DRILL-7203](https://issues.apache.org/jira/browse/DRILL-7203): Accept impersonation userName as form field & fix back button for query page

## Description

Requiring the impersonation username to be sent as an HTTP header made
it difficult to implement the Web UI submission.  This changes it so
that the userName can be sent as a form field instead, allowing the Web
UI to do a "normal" form submission, which fixes the back button, avoids
the use of the deprecated "document.open" API call, and may avoid other
glitches in the future.

## Documentation

No updates needed, this is just something "under the hood" for the Web UI.

## Testing

I added some automated tests.  I also tested manually in the Web UI.

(Please describe how this PR has been tested.)

### CODE REVIEW

**Comment by arina-ielchiieva** [comment]
@dobesv won't think break original behavior?
Please see original Jira (https://issues.apache.org/jira/browse/DRILL-5726) when making changes we should make sure original behavior is preserved since user may already be using previous approach.

**Comment by dobesv** [comment]
Which original behavior are you thinking of ?  Maybe you could write a test case to show what would be broken?

**Comment by arina-ielchiieva** [comment]
@dobesv please look into Jira description (https://issues.apache.org/jira/browse/DRILL-5726), it indicates three ways to pass user name to Drill under `Adding user name header approaches`. I am concerned if your changes did not break curl and Java ways.

**Comment by dobesv** [comment]
@arina-ielchiieva If the userName parameter is not provided, it will behave as before.  The only thing I changed related to the use of the `User-name` header is to not return an error if it not provided.

But if you are able to test it and find a regression, let me know.

File: exec/java-exec/src/main/resources/rest/query/query.ftl [hunk 1]
```diff
@@ -43,15 +43,13 @@
...
-     <div class="form-group">
-       <label for="userName">User Name</label>
-       <input type="text" size="30" name="userName" id="userName" placeholder="User Name">
-     </div>
-  </#if>
-
   <form role="form" id="queryForm" action="/query" method="POST">
+    <#if model.isOnlyImpersonationEnabled()>
       <div class="form-group">
+        <label for="userName">UserName</label>
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/resources/rest/query/query.ftl]: Please add space between words as before.
```suggestion
        <label for="userName">User Name</label>
```

File: exec/java-exec/src/test/java/org/apache/drill/exec/server/rest/TestQueryWrapper.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    } catch (UserException e) {
+      assertThat(e.getMessage(), containsString("User impersonation is not enabled"));
+    }
+  }
+
+  @Test
+  public void testImpersonationEnabledButUserNameNotProvided() throws Exception {
+    setupServer(true);
+    QueryWrapper.QueryResult result = runWithUserName("SELECT CATALOG_NAME, SCHEMA_NAME FROM information_schema.SCHEMATA", null);
+    final PersistentStore<UserBitShared.QueryProfile> profiles = workManager.getContext().getProfileStoreContext().getCompletedProfileStore();
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/server/rest/TestQueryWrapper.java]: Profiles are never used, please remove.

File: exec/java-exec/src/test/java/org/apache/drill/exec/server/rest/TestQueryWrapper.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    assertNotEquals(0, result.rows.size());
+    assertEquals(1, result.columns.size());
+    assertEquals(result.columns.iterator().next(), "SCHEMA_NAME");
+  }
+
+  @Test
+  public void testImpersonation() throws Exception {
+    setupServer(true);
+    QueryWrapper.QueryResult result = runWithUserName("SELECT CATALOG_NAME, SCHEMA_NAME FROM information_schema.SCHEMATA", "alfred");
+    final PersistentStore<UserBitShared.QueryProfile> profiles = workManager.getContext().getProfileStoreContext().getCompletedProfileStore();
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/server/rest/TestQueryWrapper.java]: Profiles are never used, please remove.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java [hunk 4]
```diff
@@ -48,13 +53,21 @@
   private final String queryType;
   private final int autoLimitRowCount;
 
+  @Nullable
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: Please remove nullable annotation.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java [hunk 5]
```diff
@@ -48,13 +53,21 @@
...
+  private final String userName;
+
   private static MemoryMXBean memMXBean = ManagementFactory.getMemoryMXBean();
 
   @JsonCreator
-  public QueryWrapper(@JsonProperty("query") String query, @JsonProperty("queryType") String queryType, @JsonProperty("autoLimit") String autoLimit) {
+  public QueryWrapper(
+    @JsonProperty("query") String query,
+    @JsonProperty("queryType") String queryType,
+    @Nullable @JsonProperty("autoLimit") String autoLimit,
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: Please remove nullable annotation.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java [hunk 6]
```diff
@@ -48,13 +53,21 @@
...
+    @Nullable @JsonProperty("userName") String userName) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java]: Please remove nullable annotation.

**Comment by arina-ielchiieva** [comment]
Restarted CI jobs, unfortunately newly added tests keep failing, though they pass on local machine, issue should be fixed since we cannot afford breaking CI with new commits.

**Comment by dobesv** [comment]
I fixed the alignment issue:

image

Thanks again for your diligent work on this review!

**Comment by dobesv** [comment]
I ran it locally using the `act` tool to simulate a github build and got a full stack trace:

```
$ act -P ubuntu-latest=nektos/act-environments-ubuntu:18.04 --job build
...
| [ERROR] org.apache.drill.exec.store.parquet.TestParquetMetadataVersion.testWrongDelimiter  Time elapsed: 0.002 s  <<< ERROR!
| java.lang.Exception: Unexpected exception, expected<org.apache.drill.common.exceptions.DrillRuntimeException> but was<java.lang.NoSuchMethodError>
|       at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:28)
|       at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
|       at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
|       at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
|       at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
|       at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
|       at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
|       at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
|       at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
|       at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
|       at org.junit.runners.Suite.runChild(Suite.java:128)
|       at org.junit.runners.Suite.runChild(Suite.java:27)
|       at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
|       at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
|       at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
|       at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
|       at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
|       at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
|       at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
|       at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
|       at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeLazy(JUnitCoreWrapper.java:119)
|       at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:87)
|       at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
|       at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
|       at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377)
|       at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138)
|       at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465)
|       at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)
| Caused by: java.lang.NoSuchMethodError: 'void org.apache.drill.common.exceptions.DrillRuntimeException.format(java.lang.String, java.lang.Object[])'
|       at org.apache.drill.exec.store.parquet.metadata.MetadataVersion.<init>(MetadataVersion.java:45)
|       at org.apache.drill.exec.store.parquet.TestParquetMetadataVersion.testWrongDelimiter(TestParquetMetadataVersion.java:96)
| 
...
```

**Comment by dobesv** [comment]
Is there any reason not to add `-DtrimStackTrace=false` to the build options in the github action?  This would allow the full stack trace to be printed in the github CI builds in the future.

**Comment by dobesv** [comment]
Looks like I fixed the issue with the builds, except that the build timed out this time.

**Comment by paul-rogers** [comment]
Cool! What was the issue? The commits are squashed so I can't just check the delta.

**Comment by dobesv** [comment]
In the CI environment I had to pull the query profile from the history but in local development the query profile was only available from the foreman.  Kind of odd.  I wonder if there's some step I could have added to make this consistent.  For now I just check one and then the other.

**Comment by arina-ielchiieva** [comment]
+1, LGTM.

**Comment by arina-ielchiieva** [comment]
@dobesv looks like you changes broke functionality which indicates query execution process.
Before your changes when running the query, you could see the following widget:

image

Example of the long running query:
```
SELECT *
FROM cp.`employee.json`
WHERE employee_id+0 < employee_id
  OR employee_id+1 < employee_id
  AND employee_id+2 < employee_id
  OR employee_id+3 < employee_id
  AND employee_id+4 < employee_id
  OR employee_id+5 < employee_id
  AND employee_id+6 < employee_id
  OR employee_id+7 < employee_id
  AND employee_id+8 < employee_id
  OR employee_id+9 < employee_id
  AND employee_id+10 < employee_id
  OR employee_id+11 < employee_id
  AND employee_id+12 < employee_id
  OR employee_id+13 < employee_id
  AND employee_id+14 < employee_id
  OR employee_id+15 < employee_id
  AND employee_id+16 < employee_id
  OR employee_id+17 < employee_id
  AND employee_id+18 < employee_id
  OR employee_id+19 < employee_id
  AND employee_id+20 < employee_id
  OR employee_id+21 < employee_id
  AND employee_id+22 < employee_id
  OR employee_id+23 < employee_id
  AND employee_id+24 < employee_id
  OR employee_id+25 < employee_id
  AND employee_id+26 < employee_id
  OR employee_id+27 < employee_id
  AND employee_id+28 < employee_id
  OR employee_id+29 < employee_id
  AND employee_id+30 < employee_id
  OR employee_id+31 < employee_id
  AND employee_id+32 < employee_id
  OR employee_id+33 < employee_id
  AND employee_id+34 < employee_id
  OR employee_id+35 < employee_id
  AND employee_id+36 < employee_id
  OR employee_id+37 < employee_id
  AND employee_id+38 < employee_id
  OR employee_id+39 < employee_id
  AND employee_id+40 < employee_id
  OR employee_id+41 < employee_id
  AND employee_id+42 < employee_id
  OR employee_id+43 < employee_id
  AND employee_id+44 < employee_id
  OR employee_id+45 < employee_id
  AND employee_id+46 < employee_id
  OR employee_id+47 < employee_id
  AND employee_id+48 < employee_id
  OR employee_id+49 < employee_id
  AND TRUE;
```
It won't return any results but will execute approximately 20 seconds which will be enough the show the functionality existed before.

**Comment by dobesv** [comment]
@arina-ielchiieva Strange, I remember that working when I tested it before, I'll take another look.

**Comment by dobesv** [comment]
@arina-ielchiieva It still shows that for me, do you have any errors in the JavaScript console?

e.g. https://youtu.be/GSCoMBPHOFc

**Comment by arina-ielchiieva** [comment]
@dobesv thanks for the video, re-checked one more time. Looks like issue is Safari specific, it works fine in Chrome though. The problem is that it worked in Safari before (I have checked on current master). Also checked fo js errors in console, there was no.

**Comment by dobesv** [comment]
Hmm Safari, that does make it a bit more tricky since I'm using Windows & Linux only currently.  If there's no error in the console ... maybe the 
```
$(window).bind("pageshow", function(event) { closePopup();});
```
is triggering too soon and closing the popup.  If you have time, maybe you could try deleting that line, or move it to the top of the function, see if that makes any difference.  Otherwise the code seems quite similar to what was there before.

**Comment by arina-ielchiieva** [comment]
@dobesv tried your suggestions, unfortunately in Safari it did not make any difference.

**Comment by arina-ielchiieva** [comment]
I think we should still merge this PR since in Drill we don't have list of browsers we support, and Chrome is the one that is commonly used, since waiting widget works in Chrome, the fact it does not work in Safari should not block this PR.

**Comment by dobesv** [comment]
@arina-ielchiieva Thanks.  Sorry I didn't have the resources to fix the Safari issue - maybe someone with macOS will have a chance to fix it in a future PR.

**Pull request closed without merging**